Make Security Agent a Windows Service

mikezhu-dd opened the pull request:
This commit makes Security Agent run as a service on Windows by mimicking the work done in dogstatsd.

### What does this PR do?

Make security agent run as a service on Windows

### Motivation

This is needed to run the agent on Windows

### Additional Notes

This commit split main.go to a NIX version and a Windows version. Main logic in app.go is changed to split the start and stop logic into RunAgent and StopAgent function. The old deferred function calls are moved to StopAgent.

This commit doesn't add the agent to installer package and doesn't add icon to the binary resource. They will be addressed in other PRs.

### Describe your test plan

Test on Windows by doing a build, copy security agent to Windows test box, run 

`New-Service -Name "datadog-security-agent" -StartupType "Manual" -BinaryPathName "C:\Program Files\Datadog\Datadog Agent\bin\agent\security-agent.exe"`

Commit:
Make Security Agent a Windows Service

This commit makes Security Agent run as a service on Windows by mimicking the work done in dogstatsd.

Commit:
Make Security Agent a Windows Service

This commit makes Security Agent run as a service on Windows by mimicking the work done in dogstatsd.

Commit:
Merge branch 'master' of github.com:DataDog/datadog-agent

derekwbrown commented on cmd/security-agent/main_windows.go:
```diff
@@ -0,0 +1,176 @@
+
+	if err != nil {
+		log.Errorf("Failed to start agent %v", err)
+		elog.Error(0xc0000008, err.Error())
```
all ofthe `elog` calls should go to the event viewer.  However, I don't see a message catalog, so all of the messages are going to just generate hard-to-understand errors in the log viewer.

derekwbrown commented on cmd/security-agent/main_windows.go:
```diff
@@ -0,0 +1,176 @@
+		changes <- svc.Status{State: svc.Stopped}
+		return
+	}
+	elog.Info(0x40000003, ServiceName)
```
For instance, this is going to just generate a `string not found` error on every startup.

Commit:
Merge branch 'master' of github.com:DataDog/datadog-agent

Commit:
Revert "Merge branch 'master' of github.com:DataDog/datadog-agent"

This reverts commit a85dc1a07ddfaaaa1fb920a131e4138a14aee6ff, reversing
changes made to 4699f0e7c6ab3d163b29e3ed68b3239b4959bcbd.

Commit:
Make Security Agent a Windows Service

This commit makes Security Agent run as a service on Windows by mimicking the work done in dogstatsd.

# Conflicts:
#	cmd/security-agent/app/app.go
#	cmd/security-agent/main_windows.go

Commit:
Merge branch 'mike.zhu/SecurityAgentWinSrv' of github.com:DataDog/datadog-agent into mike.zhu/SecurityAgentWinSrv

# Conflicts:
#	cmd/security-agent/app/app.go
#	cmd/security-agent/main_windows.go

Commit:
Fix go fmt error

Commit:
Remove extra file due to incorrect merge

brycekahle commented on pkg/ebpf/bytecode/runtime/integrity.go:
```diff
@@ -57,30 +56,12 @@ func genIntegrity(inputFile, outputFile, pkg string) error {
-		packagePrefix = "runtime."
-		imports = "import \"github.com/DataDog/datadog-agent/pkg/ebpf/bytecode/runtime\"\n"
-	}
-
```
I think you've got a bad merge because these lines should not be removed.

mikezhu-dd commented on cmd/security-agent/main_windows.go:
```diff
@@ -0,0 +1,176 @@
+func (m *myservice) Execute(args []string, r <-chan svc.ChangeRequest, changes chan<- svc.Status) (ssec bool, errno uint32) {
+	const cmdsAccepted = svc.AcceptStop | svc.AcceptShutdown
+	changes <- svc.Status{State: svc.StartPending}
+	changes <- svc.Status{State: svc.Running, Accepts: cmdsAccepted}
```
I've created a new PR because the merge from master to this branch went wrong...

The new PR is https://github.com/DataDog/datadog-agent/pull/7791

The issue has been fixed .
